src/bundle/Security/Http/Authenticator/Passport/TwoFactorPassport.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\Http\Authenticator\Passport;
+
+use Scheb\TwoFactorBundle\Security\Authentication\Exception\InvalidTwoFactorCodeException;
+use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorTokenInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\BadgeInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\Credentials\CredentialsInterface;
+use Symfony\Component\Security\Http\Authenticator\Passport\PassportInterface;
+
+class TwoFactorPassport implements PassportInterface
+{
+    /**
+     * @var TwoFactorTokenInterface
+     */
+    private $twoFactorToken;
+
+    /**
+     * @var BadgeInterface[]
+     */
+    private $badges = [];
+
+    public function __construct(TwoFactorTokenInterface $twoFactorToken, CredentialsInterface $credentials, array $badges)
+    {
+        $this->twoFactorToken = $twoFactorToken;
+        $this->addBadge($credentials);
+        foreach ($badges as $badge) {
+            $this->addBadge($badge);
+        }
+    }
+
+    public function getTwoFactorToken(): TwoFactorTokenInterface
+    {
+        return $this->twoFactorToken;
+    }
+
+    public function addBadge(BadgeInterface $badge): PassportInterface
+    {
+        $this->badges[\get_class($badge)] = $badge;
+
+        return $this;
+    }
+
+    public function hasBadge(string $badgeFqcn): bool
+    {
+        return isset($this->badges[$badgeFqcn]);
+    }
+
+    public function getBadge(string $badgeFqcn): ?BadgeInterface
+    {
+        return $this->badges[$badgeFqcn] ?? null;
+    }
+
+    public function checkIfCompletelyResolved(): void
+    {
+        foreach ($this->badges as $badge) {
+            if (!$badge->isResolved()) {
+                throw new InvalidTwoFactorCodeException('Invalid two-factor authentication code.');
+            }
+        }
+    }
```

Comment:
I wasn't sure if `PassportTrait` would be safe to use for a 3rd party. I want to avoid behavioural changes being silently introduced, so I decided to write these few lines on my on. What's your view on this?

Reply:
I intentionally didn't add `@experimental` to the trait, so it is included in the BC promise (once out of experimental state).

The way I see it is exactly the opposite: For any non-2fa related things in the passport, you want it to behave as it does for the Symfony version it's used in. Let's say Symfony changes things in the passport in 5.2. If you use the trait, Symfony would have updated it as well and the current `TwoFactorPassport` is already compatible. If you don't use the trait, you would have to update your class (and maybe even do some BC magic).

Reply:
Thanks for the clarification, using `PassportTrait` now.